src/picongpu/include/particles/ionization/byField/BSI/BSI.hpp

```diff
@@ -39,23 +50,81 @@ namespace ionization
+             * \tparam T_ParticleStorage \see picongpu/src/libPMacc/include/math/MapTuple.hpp
+             * 
+             * \param currentStep current time step */
+            template<typename T_SrcSpecies,typename T_ParticleStorage>
+            void operator()(T_SrcSpecies& src, T_ParticleStorage& pst, const uint32_t currentStep)
+            {
+
+                /* alias for electron species to be created */
+                PMACC_AUTO(electrons,pst[typename MakeIdentifier<T_DestSpecies>::type()]);
+                /* types for the source species */
+                typedef T_SrcSpecies SrcSpeciesType;
+                typedef typename SrcSpeciesType::FrameType FrameType;
+
+                DataConnector &dc = Environment<>::get().DataConnector();
+
+                fieldE = &(dc.getData<FieldE > (FieldE::getName(), true));
+                fieldB = &(dc.getData<FieldB > (FieldB::getName(), true));
+                /* specify field to particle interpolation scheme */
+                typedef typename PMacc::traits::Resolve<
+                    typename GetFlagType<FrameType,interpolation<> >::type
+                >::type InterpolationScheme;
+
+                /* margins around the supercell for the interpolation of the field on the cells */
+                typedef typename GetMargin<InterpolationScheme>::LowerMargin LowerMargin;
+                typedef typename GetMargin<InterpolationScheme>::UpperMargin UpperMargin;
+
+                /* define type of frame ionizer 
+                 * that moves over the frames with the particles and executes an operation on them */
+                typedef particles::ionization::IonizeParticlesPerFrame<IonizationAlgorithm, 
+                    MappingDesc::SuperCellSize,
+                    InterpolationScheme,
+                    fieldSolver::NumericalCellType > FrameIonizer;
+
+                /* relevant area of a block */
+                typedef SuperCellDescription<
+                    typename MappingDesc::SuperCellSize,
+                    LowerMargin,
+                    UpperMargin
+                    > BlockArea;
+
+                /* 3-dim vector : number of threads to be started in every dimension */
+                dim3 block( MappingDesc::SuperCellSize::toRT().toDim3() );
+
+                /* kernel call : instead of name<<<blocks, threads>>> (args, ...) 
+                 * "blocks" will be calculated from "this->cellDescription" and "CORE + BORDER" 
+                 * "threads" is calculated from the previously defined vector "block" */
+                __picKernelArea( particles::ionization::kernelIonizeParticles<BlockArea>, *cellDescription, CORE + BORDER )
+                    (block)
+                    ( src.getDeviceParticlesBox( ),
+                      electrons->getDeviceParticlesBox( ),
+                      fieldE->getDeviceDataBox( ),
+                      fieldB->getDeviceDataBox( ),
+                      FrameIonizer( )
+                      );
```

Comment:
What if my ADK implementation would need the kernel call to look like this: 
```c++
/* random number generator for Monte Carlo */
typedef particles::ionization::RandomNrForMonteCarlo<T_SrcSpecies> T_RandomGen;
               
T_RandomGen RandomGen(currentStep);

/* kernel call : instead of name<<<blocks, threads>>> (args, ...) 
 * "blocks" will be calculated from "this->cellDescription" and "CORE + BORDER" 
 * "threads" is calculated from the previously defined vector "block" */
__picKernelArea( particles::ionization::kernelIonizeParticles<BlockArea>, *cellDescription, CORE + BORDER )
                    (block)
                    ( src.getDeviceParticlesBox( ),
                      electrons->getDeviceParticlesBox( ),
                      fieldE->getDeviceDataBox( ),
                      fieldB->getDeviceDataBox( ),
                      FrameIonizer( ),
                      RandomGen
                      );
```
?

Do you think I should restructure the implementation more or is there an elegant solution to handle that the kernel sometimes gets a random generator as an argument and sometimes not?